### ISSUE TITLE
Memory Leaks when using most -listxxxx commands

### ISSUE THREAD

**Issue by Tafoid** [mamedev/mame#325]
-listmedia/slot/device/software/xml (at least)
-- memory leak warning --
Error message spam (nofree) pointing at:
"../../../../../src/emu/netlist/plib/plists.h"
"../../../../../src/emu/netlist/plib/pstring.c"
As show in http://pastebin.com/YRTd47zL

Appears to have broken with this submit on August 27, 2015
https://github.com/mamedev/mame/commit/bce5387096ff117dac9bc2409d77fb1ad77528c1

**Comment by LnmVolbo** [comment]
Trying out a number of different games with the -listxml command, only a few show the memory leak.  So far, I have found three source files that cause the error:

1942.c (1942 and clones)
m62.c (battroad, etc.)
pong.c (breakout, pong & clones)

I'm still parsing the list of games, only completed from numbers to the end of letter b.  I will update this list after checking the rest of the games from the -listfull command.

To quickly see the affected items, without going through the entire list, you can type 

mame -listxml <game name>

where <game name> is one of the affected games listed above.

**Comment by Tafoid** [comment]
Yes.  Any sets that use the code in question involves netlist discrete emulation in some form.

**Comment by couriersud** [comment]
I will have a deeper look later this week if I don't track the issue in
the next 30 minutes.

Thanks for spotting this.

**Comment by LnmVolbo** [comment]
I noticed that not all the netlist games have the issue.

Here is a link to a spreadsheet with all the netlist games listed with a tally of their device_ref names:

https://www.sendspace.com/file/w08qh4 (14 KB)

It appears that the games only using netlist_cpu (by itself) are unaffected.

The games that are leaking use three or more of the following device_ref names:

netlist_sound
netlist_stream_input
netlist_stream_output
netlist_logic_input
netlist_analog_input
netlist_analog_output

Apparently netlist_cpu doesn't utilize any pstrings, or else it uses pstrings but doesn't misuse them so that they leak.

All the leaky netlist games have an input and an output specified.  Is it possible that once committed to an input or output operation, the pstring cannot be disposed of in the expected way, and it leaks (stays in memory) instead?

Would the pstring need to be copied to a temporary value, and that temporary value committed to the I/O operation, allowing the original pstring to be discarded properly?

That sounds kind of complicated.  I gather the pstrings were added to speed up string operations.  Duplicating them for I/O might negate the speed increase.

The pstring code pertaining to allocation and freeing is a bit terse for me to follow, so I am unable to make any more specific suggestions about solving this issue.

-- I could probably come up with some more goofy questions tho.
